File: wrench-app/src/main/java/com/izettle/wrench/dialogs/enumvalue/PredefinedValueRecyclerViewAdapter.java
package com.izettle.wrench.dialogs.enumvalue;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.izettle.wrench.database.WrenchPredefinedConfigurationValue;
import com.izettle.wrench.databinding.SimpleListItemBinding;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import androidx.recyclerview.widget.RecyclerView;

public class PredefinedValueRecyclerViewAdapter extends RecyclerView.Adapter<PredefinedValueRecyclerViewAdapter.ViewHolder> {

    private final ArrayList<WrenchPredefinedConfigurationValue> items = new ArrayList<>();
    private final Listener listener;

    PredefinedValueRecyclerViewAdapter(Listener listener) {
        this.listener = listener;
    }

    @NotNull
    @Override
    public ViewHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
        SimpleListItemBinding binding = SimpleListItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NotNull ViewHolder holder, int position) {
        WrenchPredefinedConfigurationValue predefinedConfigurationValue = items.get(position);

        holder.binding.value.setText(predefinedConfigurationValue.getValue());
        holder.binding.getRoot().setOnClickListener(listener::onClick);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    void setItems(List<WrenchPredefinedConfigurationValue> items) {
        this.items.clear();
        this.items.addAll(items);
        notifyDataSetChanged();
    }

    WrenchPredefinedConfigurationValue getItem(int position) {
        return items.get(position);
    }

    interface Listener {
        void onClick(View view);
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private final SimpleListItemBinding binding;

        ViewHolder(SimpleListItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: wrench-app/src/main/java/com/izettle/wrench/dialogs/enumvalue/PredefinedValueRecyclerViewAdapter.kt
package com.izettle.wrench.dialogs.enumvalue

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.izettle.wrench.database.WrenchPredefinedConfigurationValue
import com.izettle.wrench.databinding.SimpleListItemBinding

class PredefinedValueRecyclerViewAdapter internal constructor(
        private val listener: Listener
) : ListAdapter<WrenchPredefinedConfigurationValue, PredefinedValueRecyclerViewAdapter.ViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = SimpleListItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = getItem(position)

        holder.binding.value.text = item.value
        holder.binding.root.setOnClickListener { listener.onClick(it, item) }
    }

    internal interface Listener {
        fun onClick(view: View, item: WrenchPredefinedConfigurationValue)
    }

    inner class ViewHolder(internal val binding: SimpleListItemBinding) : RecyclerView.ViewHolder(binding.root)

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<WrenchPredefinedConfigurationValue>() {
            override fun areItemsTheSame(oldApplication: WrenchPredefinedConfigurationValue, newApplication: WrenchPredefinedConfigurationValue): Boolean {
                return oldApplication.id == newApplication.id
            }

            override fun areContentsTheSame(oldApplication: WrenchPredefinedConfigurationValue, newApplication: WrenchPredefinedConfigurationValue): Boolean {
                return oldApplication == newApplication
            }
        }
    }
}
